WIP: config rewrite

Rick-Anderson opened the pull request.

Rick-Anderson commented:
@HaoK  are you the right person to review this? Do you want me to send you a word doc? cc @tdykstra

Rick-Anderson commented:
Word version at https://www.dropbox.com/s/ujk0jk5i27kfywo/Configuration.docx?dl=0 if that's easier.  @HaoK  can you suggestion someone to do a quick review?

HaoK commented:
@divega @glennc should also review, I will take a look as well

Rick-Anderson commented:
@HaoK  @divega @glennc can you give me an ETA for the review?

Word version at https://www.dropbox.com/s/ujk0jk5i27kfywo/Configuration.docx?dl=0 if that's easier.  I could also send you the HTML version.

HaoK commented:
You should update the startup code in general to use the new pattern:

```
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
```

instead of Directory.GetCurrentPath() to be consistent with the templates

HaoK commented:
I did a first pass, and commented some in the word doc via dropbox, you should get @divega to do a final review as well

tdykstra commented on aspnet/fundamentals/configuration.rst:
```diff
@@ -2,169 +2,252 @@
+.. _custom-config-providers:
 
-You configure options using the :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>` extension method. You can configure options using a delegate or by binding your options to configuration:
+Note that Dependency Injection (DI) is not setup until after ``ConfigureServices`` is invoked and the configuration system is not DI aware.
```
Not clear why this note is here as the issue is not directly relevant to this particular example.  Also the sentence structure is ambiguous, and it's not clear what "not DI aware" means.

Rick-Anderson replied:
moved to bottom.  This is a copy/paste from Dan.Roth, folks that need to know will figure it out.

tdykstra commented on aspnet/fundamentals/configuration.rst:
```diff
@@ -2,169 +2,252 @@
+In this section we'll create a simple configuration provider that reads name-value pairs from a database using EF.
 
-To start off we'll define a simple ``ConfigurationValue`` entity for storing configuration values in the database:
+Define a `ConfigurationValue`` entity for storing configuration values in the database:
```
extra backtick

Rick-Anderson commented:
@HaoK  wrote:

You should update the startup code in general to use the new pattern:

```
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
```

instead of Directory.GetCurrentPath() to be consistent with the templates

I'm guessing you meant `Directory.GetCurrentDirectory()`, not `Directory.GetCurrentPath()`
Do you remember where you saw this? I only have two Web projects, and they both use the pattern in the RTM templates.

Rick-Anderson commented on aspnet/fundamentals/configuration.rst:
```diff
@@ -2,169 +2,252 @@
+.. _options-example:
 
-.. warning:: You should never store passwords or other sensitive data in configuration provider code or in plain text configuration files. You also shouldn't use production secrets in your development or test environments. Instead, such secrets should be specified outside the project tree, so they cannot be accidentally committed into the configuration provider repository. Learn more about :doc:` ...
+In the code below, ``ConfigurationBuilder`` is initialized in the ``Startup`` class. In the ``ConfigureServices`` method, an ``IConfigurationRoot`` interface is passed to the to :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>`, which adds ``ConfigurationBuilder`` to the dependency injection container:
```
Per the style guide I'm removing the description and moving all code explanations to code comments.

>  Why do we use IConfigurationRoot here but we used IConfiguration in the first snippet in the article?

Nice catch. It should be IConfigurationRoot  everywhere, we want the root of the configuration.  IConfiguration is the lower level interface that is specialized by `IConfigurationRoot` and `IConfigurationSection`

Rick-Anderson commented on aspnet/fundamentals/configuration.rst:
```diff
@@ -2,169 +2,252 @@
+There is no limit to the number of ``IConfigureOptions<TOptions>`` services you can add. Each configuration service comes in a NuGet package. They are all applied in order they are registered. Each call to :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>` adds an :dn:iface:`~Microsoft.Extensions.Options.IConfigureOptions\<TOptions>` ser ...
 
-A simple ``MyOptions`` class is shown here:
+.. note:: Configuration keys are case insensitive.
```
Moved it up, removed note. Remove 2 notes, 1 tip and 1 warning all to a bullet list.

Rick-Anderson commented:
@HaoK per your DB feedback (I fixed all  your suggestions):

- Spit out `GetValue `in it's own section
- Explained `GetValue` is basically just syntax sugar for `GetSection(<key>).TypeConverter.Convert<T>` and it's not doing any binding
- Added ctor to models and explained that's normally how you set defaults.

@divega @glennc can you review? I can send you a word file if you like.  I'll need a shipit

Rick-Anderson commented:
@HaoK  I added:

The built in configuration providers are not refreshed when the configuration data changes. If your configuration data changes, you'll need to restart your app to get the new data.

Now how do I explain `reloadOnChange`?
```
public static IConfigurationBuilder AddJsonFile(this IConfigurationBuilder builder,
                                           string path, bool optional, bool reloadOnChange)

```

Rick-Anderson commented:
see https://github.com/aspnet/Configuration/issues/532

Rick-Anderson commented:
moved to #2120  -
